### DISCUSSION TITLE
[Proposal] Named parameters or warning about unnamed parameter matching name but out of place

### DISCUSSION THREAD

**Discussion by tymtam2** [dotnet/csharplang#6318] [Language Ideas]
Would it be possible to get some help with writing a proposal for the 2 warnings described below? 

I looked through this repo and I could not find propsals which talk about generating warnings/errors. What's the etiquette? Once I understand this I can rewrite into specish.

----

# Warning/requirement 1: 

## Summary

Introduce a warning (or error) for when an unnamed argument is called out of place. 

```
string a = "over";
string some_name = "take";

F(some_name, a); // Warning/error: An unnamed argument that matches argument name at different position found. Please use a named argument.
F(some_name, b: a); // OK

string F(string a, string b) => a+b;
```

# Motivation

Eliminate hard-to-spot mistakes. 

# Warning/requirement 2: 

## Summary 

Introduce an opt-in option for disallowing unnamed arguments, unless the argument name matches at the exact position.

```
string a = "over";
string b = "take";
string some_name = "123";

F("A", "B"); // Not allowed - not named and not matching
F(b, a); // Not allowed - not named and not matching
F(a, some_name); // Not allowed - 'a' matches by name and place , but b not named 
F(some_name, a); // Not allowed - 1) unnamed and doesn't match name  & 2) a matches but is not in place
F(some_name, some_name); // Not allowed - 1) 2x unnamed and doesn't match name


F(a, b); // OK - both match and are in place
F(a, b: "B"); // OK - 'a' matches name and position, b is named
F(a: "A", b: "B"); // OK - both named
F(a, b: some_name); // OK - both named
F(b: a, a: b); // OK - swapped position but both named
F(a: b, b: a); // OK - both named
F(b: "A", a: "A"); // OK - both named
F(a: some_name, b: a); // OK - both named

string F(string a, string b) => a+b;
```

## Motivation

Eliminate issues resulting from out of order parameters, while 
* allowing succinct code when names match
* leaving the flexibility to do swaps `F(a: b, b: a);`

```
class Service 
{
...
void Get(string userId, string targetId) => repo.Get(userId, targetId); // Correct and succinct 

void Get(string userId, string targetId) => repo.Get(userId: userId, targetId: targetId); // Correct, but verbose

void Get(string userId, string targetId) => repo.Get(targetId, userId); // Incorrect, and caught
```

**Comment by HaloFour** [comment]
You can use analyzers to accomplish both of these in your own projects.

**Reply by tymtam2** [reply]
But I want to save the whole world, not just myself. 

Seriously though, is the something I should/could read to educate myself why C# doesn't, even optionally, enforce named parameters?

**Reply by HaloFour** [reply]
> Seriously though, is the something I should/could read to educate myself why C# doesn't, even optionally, enforce named parameters?

C#, like C++ and C, is primarily a positional parameter language.  C# didn't even support named parameters until 3.0, and at that time it was added to support optional parameters.

> But I want to save the whole world, not just myself.

I think it's very unlikely for the language itself to take a position here regarding named parameters.  An analyzer is the better route, and if you think that such an analyzer would be best enforced by default you could take that to the [Roslyn Analyzers repository](https://github.com/dotnet/roslyn-analyzers) where they work on the analyzers that ship by default.

**Reply by PathogenDavid** [reply]
This might be a better proposal for [roslyn-analyzers](https://github.com/dotnet/roslyn-analyzers) instead. (Whops, HaloFour beat me to it.)

Personally in my own code I try to avoid situations where this is a problem in the first place. For example, if you use a strongly typed `UserId` type instead of `string` for user IDs this mistake becomes impossible.

**Reply by theunrepentantgeek** [reply]
> Seriously though, is the something I should/could read to educate myself why C# doesn't, even optionally, enforce named parameters?

**Why doesn't the feature exist?**

Because no-one on the LDM has ever thought this to be of sufficient value to champion the issue and get the rest of the language team onside to invest the time and resources to implement and test.

As Eric Lippert (former C# compiler team member) has [commented](https://stackoverflow.com/a/8673015):

> The question is "why does C# not have this feature?" The answer to that question is always the same. Features are unimplemented by default; C# does not have that feature because no one designed, implemented and shipped the feature to customers.

**Why not implement this as a compiler flag?**

Every additional compiler flag _doubles_ the number of variants of the language - 5 flags = 32 variants of C# to test, 10 flags = 1024 variants of C# to test, and so on. The LDM has been very clear that they would only consider exceptional circumstances involving tremendous benefit before a new flag would be added (such as the benefit of nullable-reference types). They also consider the flag controlling **checked** to be a mistake.

**Alternatives**

> But I want to save the whole world, not just myself.

If you want to have the noise of having every single parameter name shown proactively in your code, there are ways you can do this in your IDE of choice without inflicting anything on anyone else. In Rider, the option is available by default. In Visual Studio, try Resharper (if it's not already there). And so on.

I have a strong opinion about this. Personally, I can't stand the noise of having all those parameter names showing - it makes the code virtually unreadable. To me, it literally conveys exactly zero information at the considerable expense of littering the screen with debris, in much the same way (and to equally little impact) as requiring `this.` to qualify every member access. I feel it's far better to design the API in a way that it can't be misused; relying on parameter naming should be a seldom used last resort, not the technique of first choice.

If you want to have this, then go for it. Nobody is stopping you from doing this to yourself - in fact, it can already be done. But please don't assume that your opinion is universally accepted.

**Reply by tymtam2** [reply]
Thanks for the replies; it's much appreciated that you took time to answer.

**Reply by tymtam2** [reply]
Now I understand: 
1. What I want:
    a. Never have to wonder what the null means in `Get(id, null);` (my preferred solution `Get(id, atUtc: null);`)
    b. Never have to wonder if `Set(oldX, newX)` should have been `Set(newX, oldX)` (my preferred solution `Set(oldValue: oldX, newValue: newX);`) - I understand that this is subjective and a test should catch it and it takes time to get used to (which is in big part a UI issue).
2. That this is what \*I\* feel is useful, and not a universally held view.
3. That there is a tool (an analyzer) to achieve all that I want and:
   a. it can work the way I want;
   b. I can do it
   c. can be published and others can use it, if they want. 
   
Cheers again for the comments.

**Reply by yaakov-h** [reply]
For 1(a), I've written [an analyzer](https://github.com/WiseTechGlobal/WTG.Analyzers/blob/master/src/WTG.Analyzers/Analyzers/BooleanLiteralArgument/BooleanLiteralArgumentAnalyzer.cs) that does this for `Get(id, true)` and `Get(id, false)`. It shouldn't be hard to adapt this for your particular requirement of null literals instead.

**Reply by RikkiGibson** [reply]
I think it would be worth starting an issue in roslyn-analyzers to see if there's interest in taking some version of this as a disabled-by-default or info-by-default rule. If not, there are several prominent third-party Roslyn analyzer packages which might be interested in taking this rule, or might even already have some version of this rule.
